File: FBControlCore/Crashes/FBCrashLog.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBCrashLog.h"

#import <stdio.h>

#import "FBControlCoreGlobalConfiguration.h"
#import "FBConcurrentCollectionOperations.h"
#import "NSPredicate+FBControlCore.h"
#import "FBControlCoreError.h"
#import "FBControlCoreLogger.h"
#import "FBConcatedJsonParser.h"
#import "FBCrashLogParser.h"
#import <Foundation/Foundation.h>

@implementation FBCrashLog


+ (NSDateFormatter *)dateFormatter
{
  static dispatch_once_t onceToken;
  static NSDateFormatter *dateFormatter = nil;
  dispatch_once(&onceToken, ^{
    dateFormatter = [NSDateFormatter new];
    dateFormatter.dateFormat = @"yyyy-MM-dd HH:mm:ss.SSS Z";
    dateFormatter.lenient = YES;
    dateFormatter.locale = [NSLocale localeWithLocaleIdentifier:@"en_US"];
  });
  return dateFormatter;
}

#pragma mark Initializers

- (instancetype)initWithInfo:(FBCrashLogInfo *)info contents:(NSString *)contents
{
  self = [super init];
  if (!self) {
      return nil;
  }

  _info = info;
  _contents = contents;

  return self;
}

- (NSString *)description
{
  return [NSString stringWithFormat:@"Crash Info: %@ \n Crash Report: %@\n", _info, _contents];
}

#pragma mark NSCopying

- (instancetype)copyWithZone:(NSZone *)zone
{
    // Is immutable
    return self;
}

@end

@implementation FBCrashLogInfo

#pragma mark Initializers

+ (nullable instancetype)fromCrashLogAtPath:(NSString *)crashPath error:(NSError **)error {
  if (!crashPath) {
    return [[FBControlCoreError
      describe:@"No crash path provided"]
      fail:error];
  }
  if (![NSFileManager.defaultManager fileExistsAtPath:crashPath]) {
    return [[FBControlCoreError
      describeFormat:@"File does not exist at given crash path: %@", crashPath]
      fail:error];
  }
  if (![NSFileManager.defaultManager isReadableFileAtPath:crashPath]) {
    return [[FBControlCoreError
      describeFormat:@"Crash file at %@ is not readable", crashPath]
      fail:error];
  }
  NSData *crashFileData = [NSData dataWithContentsOfFile:crashPath options:0 error:error];
  if (!crashFileData) {
    return [[FBControlCoreError
      describeFormat:@"Could not read data from %@", crashPath]
      fail:error];
  } else if (crashFileData.length == 0) {
      return [[FBControlCoreError
        describeFormat:@"Crash file at %@ is empty", crashPath]
        fail:error];
  }

  NSString *crashString = [[NSString alloc] initWithData:crashFileData encoding:NSUTF8StringEncoding];
  if (!crashString) {
    return [[FBControlCoreError
      describeFormat:@"Could not extract string from %@", crashPath]
      fail:error];
  }

  return [self fromCrashLogString:crashString crashPath:crashPath parser:[self getPreferredCrashLogParserForCrashString:crashString] error:error];
}

+ (id<FBCrashLogParser>)getPreferredCrashLogParserForCrashString:(NSString *)crashString {
  if (crashString.length > 0 && [crashString characterAtIndex:0] == '{') {
    return [[FBConcatedJSONCrashLogParser alloc] init];
  } else {
    return [[FBPlainTextCrashLogParser alloc] init];
  }
}

+ (nullable instancetype)fromCrashLogString:(NSString *)crashString crashPath:(NSString *)crashPath parser:(id<FBCrashLogParser>)parser error:(NSError **)error {

  NSString *executablePath = nil;
  NSString *identifier = nil;
  NSString *processName = nil;
  NSString *parentProcessName = nil;
  NSDate *date = nil;
  pid_t processIdentifier = -1;
  pid_t parentProcessIdentifier = -1;
  NSString *exceptionDescription = nil;
  NSString *crashedThreadDescription = nil;

  NSError *err;
  [parser parseCrashLogFromString:crashString
    executablePathOut:&executablePath
    identifierOut:&identifier
    processNameOut:&processName
    parentProcessNameOut:&parentProcessName
    processIdentifierOut:&processIdentifier
    parentProcessIdentifierOut:&parentProcessIdentifier
    dateOut:&date
    exceptionDescription:&exceptionDescription
    crashedThreadDescription:&crashedThreadDescription
    error:&err];

  if (err) {
    return [[FBControlCoreError
             describeFormat:@"Could not parse crash string %@", err]
            fail:error];
  }

  if (processName == nil) {
    return [[FBControlCoreError
             describe:@"Missing process name in crash log"]
            fail:error];
  }
  if (identifier == nil) {
    return [[FBControlCoreError
             describe:@"Missing identifier in crash log"]
            fail:error];
  }
  if (parentProcessName == nil) {
    return [[FBControlCoreError
             describe:@"Missing process name in crash log"]
            fail:error];
  }
  if (executablePath == nil) {
    return [[FBControlCoreError
             describe:@"Missing executable path in crash log"]
            fail:error];
  }
  if (processIdentifier == -1) {
    return [[FBControlCoreError
             describe:@"Missing process identifier in crash log"]
            fail:error];
  }
  if (parentProcessIdentifier == -1) {
    return [[FBControlCoreError
             describe:@"Missing parent process identifier in crash log"]
            fail:error];
  }
  if (date == nil) {
    return [[FBControlCoreError
             describe:@"Missing date in crash log"]
            fail:error];
  }

  FBCrashLogInfoProcessType processType = [self processTypeForExecutablePath:executablePath];

  return [[FBCrashLogInfo alloc]
    initWithCrashPath:crashPath
    executablePath:executablePath
    identifier:identifier
    processName:processName
    processIdentifier:processIdentifier
    parentProcessName:parentProcessName
    parentProcessIdentifier:parentProcessIdentifier
    date:date
    processType:processType
    exceptionDescription:exceptionDescription
    crashedThreadDescription:crashedThreadDescription];
}


- (instancetype)initWithCrashPath:(NSString *)crashPath executablePath:(NSString *)executablePath identifier:(NSString *)identifier processName:(NSString *)processName processIdentifier:(pid_t)processIdentifer parentProcessName:(NSString *)parentProcessName parentProcessIdentifier:(pid_t)parentProcessIdentifier date:(NSDate *)date processType:(FBCrashLogInfoProcessType)processType exceptionDescription:(NSString *)exceptionDescription crashedThreadDescription:(NSString *)crashedThreadDescription{
  self = [super init];
  if (!self) {
    return nil;
  }

  _crashPath = crashPath;
  _executablePath = executablePath;
  _identifier = identifier;
  _processName = processName;
  _processIdentifier = processIdentifer;
  _parentProcessName = parentProcessName;
  _parentProcessIdentifier = parentProcessIdentifier;
  _date = date;
  _processType = processType;
  _exceptionDescription = exceptionDescription;
  _crashedThreadDescription = crashedThreadDescription;

  return self;
}

#pragma mark Public

+ (BOOL)isParsableCrashLog:(NSData *)data
{
#if defined(__apple_build_version__)
  NSString *crashString = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
  FBCrashLogInfo *parsable = [self fromCrashLogString:crashString crashPath:@"" parser:[self getPreferredCrashLogParserForCrashString:crashString] error:nil];
  return parsable != nil;
#else
  return NO;
#endif
}

#pragma mark NSObject

- (NSString *)description
{
  return [NSString stringWithFormat:
    @"Identifier %@ | Executable Path %@ | Process %@ | pid %d | Parent %@ | ppid %d | Date %@ | Path %@ | Exception: %@ | Trace: %@",
    self.identifier,
    self.executablePath,
    self.processName,
    self.processIdentifier,
    self.parentProcessName,
    self.parentProcessIdentifier,
    self.date,
    self.crashPath,
    self.exceptionDescription,
    self.crashedThreadDescription
  ];
}

#pragma mark NSCopying

- (instancetype)copyWithZone:(NSZone *)zone
{
  // Is immutable
  return self;
}

#pragma mark Properties

- (NSString *)name
{
  return self.crashPath.lastPathComponent;
}

- (nullable NSString *)loadRawCrashLogStringWithError:(NSError **)error;
{
  return [NSString stringWithContentsOfFile:self.crashPath encoding:NSUTF8StringEncoding error:error];
}

#pragma mark Bulk Collection

+ (NSArray<FBCrashLogInfo *> *)crashInfoAfterDate:(NSDate *)date logger:(id<FBControlCoreLogger>)logger
{
  NSMutableArray<FBCrashLogInfo *> *allCrashInfos = NSMutableArray.new;

  for (NSString *basePath in self.diagnosticReportsPaths) {
    NSArray<FBCrashLogInfo *> *crashInfos = [[FBConcurrentCollectionOperations
      filterMap:[NSFileManager.defaultManager contentsOfDirectoryAtPath:basePath error:nil]
      predicate:[FBCrashLogInfo predicateForFilesWithBasePath:basePath afterDate:date withExtension:@"crash"]
      map:^ FBCrashLogInfo * (NSString *fileName) {
        NSString *path = [basePath stringByAppendingPathComponent:fileName];
        NSError *error = nil;
        FBCrashLogInfo *info = [FBCrashLogInfo fromCrashLogAtPath:path error:&error];
        if (!info) {
          [logger logFormat:@"Error parsing log %@", error];
        }
        return info;
      }]
      filteredArrayUsingPredicate:NSPredicate.notNullPredicate];

    [allCrashInfos addObjectsFromArray:crashInfos];
  }

  return [allCrashInfos copy];
}

#pragma mark Contents

- (FBCrashLog *)obtainCrashLogWithError:(NSError **)error
{
  NSError *innerError = nil;
  NSString *contents = [self loadRawCrashLogStringWithError:&innerError];
  if (!contents) {
    return [[[FBControlCoreError
      describeFormat:@"Failed to read crash log at path %@", self.crashPath]
      causedBy:innerError]
      fail:error];
  }
  return [[FBCrashLog alloc] initWithInfo:self contents:contents];
}

#pragma mark Predicates

+ (NSPredicate *)predicateForCrashLogsWithProcessID:(pid_t)processID
{
  return [NSPredicate predicateWithBlock:^ BOOL (FBCrashLogInfo *crashLog, id _) {
    return crashLog.processIdentifier == processID;
  }];
}

+ (NSPredicate *)predicateNewerThanDate:(NSDate *)date
{
  return [NSPredicate predicateWithBlock:^ BOOL (FBCrashLogInfo *crashLog, id _) {
    return [date compare:crashLog.date] == NSOrderedAscending;
  }];
}

+ (NSPredicate *)predicateOlderThanDate:(NSDate *)date
{
  return [NSCompoundPredicate notPredicateWithSubpredicate:[self predicateNewerThanDate:date]];
}

+ (NSPredicate *)predicateForIdentifier:(NSString *)identifier
{
  return [NSPredicate predicateWithBlock:^ BOOL (FBCrashLogInfo *crashLog, id _) {
    return [identifier isEqualToString:crashLog.identifier];
  }];
}

+ (NSPredicate *)predicateForName:(NSString *)name
{
  return [NSPredicate predicateWithBlock:^ BOOL (FBCrashLogInfo *crashLog, id _) {
    return [name isEqualToString:crashLog.name];
  }];
}

+ (NSPredicate *)predicateForExecutablePathContains:(NSString *)contains
{
  return [NSPredicate predicateWithBlock:^ BOOL (FBCrashLogInfo *crashLog, id _) {
    return [crashLog.executablePath containsString:contains];
  }];
}

#pragma mark Helpers

+ (NSArray<NSString *> *)diagnosticReportsPaths
{
  return @[
    [NSHomeDirectory() stringByAppendingPathComponent:@"Library/Logs/DiagnosticReports"],
    @"/Library/Logs/DiagnosticReports", // diagnostic reports path when ReportCrash is running as root.
  ];
}

#pragma mark Private

+ (FBCrashLogInfoProcessType)processTypeForExecutablePath:(NSString *)executablePath
{
  if ([executablePath containsString:@"Platforms/iPhoneSimulator.platform"]) {
    return FBCrashLogInfoProcessTypeSystem;
  }
  if ([executablePath containsString:@".app"]) {
    return FBCrashLogInfoProcessTypeApplication;
  }
  return FBCrashLogInfoProcessTypeCustom;
}

+ (NSPredicate *)predicateForFilesWithBasePath:(NSString *)basePath afterDate:(NSDate *)date withExtension:(NSString *)extension
{
  NSFileManager *fileManager = NSFileManager.defaultManager;
  NSPredicate *datePredicate = [NSPredicate predicateWithValue:YES];
  if (date) {
    datePredicate = [NSPredicate predicateWithBlock:^ BOOL (NSString *fileName, NSDictionary *_) {
      NSString *path = [basePath stringByAppendingPathComponent:fileName];
      NSDictionary *attributes = [fileManager attributesOfItemAtPath:path error:nil];
      return [attributes.fileModificationDate compare:date] != NSOrderedAscending;
    }];
  }
  return [NSCompoundPredicate andPredicateWithSubpredicates:@[
    [NSPredicate predicateWithFormat:@"pathExtension == %@", extension],
    datePredicate
  ]];
}

@end

File: FBControlCore/Crashes/FBCrashLog.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation

/// The kind of process that crashed.
public struct CrashLogInfoProcessType: OptionSet, Sendable {
  public let rawValue: UInt

  public init(rawValue: UInt) {
    self.rawValue = rawValue
  }

  /// A process that is part of the operating system runtime.
  public static let system = CrashLogInfoProcessType(rawValue: 1 << 0)
  /// A process that is an application.
  public static let application = CrashLogInfoProcessType(rawValue: 1 << 1)
  /// A process that is neither an application nor part of the operating system runtime.
  public static let custom = CrashLogInfoProcessType(rawValue: 1 << 2)
}

public final class FBCrashLog: CustomStringConvertible {

  public let info: CrashLogInfo
  public let contents: String

  public init(info: CrashLogInfo, contents: String) {
    self.info = info
    self.contents = contents
  }

  public var description: String {
    "Crash Info: \(info) \n Crash Report: \(contents)\n"
  }

  // MARK: - Public

  public class func dateFormatter() -> DateFormatter {
    FBCrashLog_dateFormatter
  }
}

private let FBCrashLog_dateFormatter: DateFormatter = {
  let formatter = DateFormatter()
  formatter.dateFormat = "yyyy-MM-dd HH:mm:ss.SSS Z"
  formatter.isLenient = true
  formatter.locale = Locale(identifier: "en_US")
  return formatter
}()

enum CrashLogError: Error {
  case fileDoesNotExist(path: String)
  case fileNotReadable(path: String)
  case dataReadFailed(path: String, underlying: Error)
  case fileEmpty(path: String)
  case stringExtractionFailed(path: String)
  case readFailed(path: String, underlying: Error)
  case parseFailed(underlying: Error)
  case missingField(field: String)
}

extension CrashLogError: LocalizedError {
  public var errorDescription: String? {
    switch self {
    case let .fileDoesNotExist(path):
      return "File does not exist at given crash path: \(path)"
    case let .fileNotReadable(path):
      return "Crash file at \(path) is not readable"
    case let .dataReadFailed(path, _):
      return "Could not read data from \(path)"
    case let .fileEmpty(path):
      return "Crash file at \(path) is empty"
    case let .stringExtractionFailed(path):
      return "Could not extract string from \(path)"
    case let .readFailed(path, _):
      return "Failed to read crash log at path \(path)"
    case let .parseFailed(underlying):
      return "Could not parse crash string \(underlying)"
    case let .missingField(field):
      return "Missing \(field) in crash log"
    }
  }
}

public final class CrashLogInfo: CustomStringConvertible {

  // MARK: - Properties

  public let crashPath: String
  public let executablePath: String
  public let identifier: String
  public let processName: String
  public let processIdentifier: pid_t
  public let parentProcessName: String
  public let parentProcessIdentifier: pid_t
  public let date: Date
  public let processType: CrashLogInfoProcessType
  public let exceptionDescription: String?
  public let crashedThreadDescription: String?

  public var name: String {
    (crashPath as NSString).lastPathComponent
  }

  public init(
    crashPath: String,
    executablePath: String,
    identifier: String,
    processName: String,
    processIdentifier: pid_t,
    parentProcessName: String,
    parentProcessIdentifier: pid_t,
    date: Date,
    processType: CrashLogInfoProcessType,
    exceptionDescription: String?,
    crashedThreadDescription: String?
  ) {
    self.crashPath = crashPath
    self.executablePath = executablePath
    self.identifier = identifier
    self.processName = processName
    self.processIdentifier = processIdentifier
    self.parentProcessName = parentProcessName
    self.parentProcessIdentifier = parentProcessIdentifier
    self.date = date
    self.processType = processType
    self.exceptionDescription = exceptionDescription
    self.crashedThreadDescription = crashedThreadDescription
  }

  // MARK: - Factory Methods

  public class func fromCrashLog(atPath crashPath: String) throws -> CrashLogInfo {
    let fileManager = FileManager.default
    if !fileManager.fileExists(atPath: crashPath) {
      throw CrashLogError.fileDoesNotExist(path: crashPath)
    }
    if !fileManager.isReadableFile(atPath: crashPath) {
      throw CrashLogError.fileNotReadable(path: crashPath)
    }
    let crashFileData: Data
    do {
      crashFileData = try Data(contentsOf: URL(fileURLWithPath: crashPath))
    } catch {
      throw CrashLogError.dataReadFailed(path: crashPath, underlying: error)
    }
    if crashFileData.isEmpty {
      throw CrashLogError.fileEmpty(path: crashPath)
    }

    guard let crashString = String(data: crashFileData, encoding: .utf8) else {
      throw CrashLogError.stringExtractionFailed(path: crashPath)
    }

    let parser = getPreferredCrashLogParser(forCrashString: crashString)
    return try fromCrashLogString(crashString, crashPath: crashPath, parser: parser)
  }

  public class func isParsableCrashLog(_ data: Data) -> Bool {
    #if canImport(Darwin)
    guard let crashString = String(data: data, encoding: .utf8) else {
      return false
    }
    let parser = getPreferredCrashLogParser(forCrashString: crashString)
    do {
      _ = try fromCrashLogString(crashString, crashPath: "", parser: parser)
      return true
    } catch {
      return false
    }
    #else
    return false
    #endif
  }

  public var description: String {
    "Identifier \(identifier) | Executable Path \(executablePath) | Process \(processName) | pid \(processIdentifier) | Parent \(parentProcessName) | ppid \(parentProcessIdentifier) | Date \(date) | Path \(crashPath) | Exception: \(exceptionDescription ?? "nil") | Trace: \(crashedThreadDescription ?? "nil")"
  }

  public func loadRawCrashLogString() throws -> String {
    try String(contentsOfFile: crashPath, encoding: .utf8)
  }

  // MARK: - Bulk Collection

  public class func crashInfo(afterDate date: Date, logger: FBControlCoreLogger?) -> [CrashLogInfo] {
    var allCrashInfos: [CrashLogInfo] = []

    for basePath in diagnosticReportsPaths {
      let fileNames = (try? FileManager.default.contentsOfDirectory(atPath: basePath)) ?? []
      let predicate = predicateForFiles(withBasePath: basePath, afterDate: date, withExtensions: ["crash", "ips"])
      let crashInfos = ConcurrentCollectionOperations.filterMap(
        fileNames as [Any],
        predicate: predicate,
        map: { item -> Any in
          guard let fileName = item as? String else {
            return NSNull()
          }
          let path = (basePath as NSString).appendingPathComponent(fileName)
          do {
            return try CrashLogInfo.fromCrashLog(atPath: path)
          } catch {
            logger?.log("Error parsing log \(error)")
            return NSNull()
          }
        }
      )
      allCrashInfos.append(contentsOf: crashInfos.compactMap { $0 as? CrashLogInfo })
    }

    return allCrashInfos
  }

  // MARK: - Contents

  public func obtainCrashLog() throws -> FBCrashLog {
    let contents: String
    do {
      contents = try loadRawCrashLogString()
    } catch {
      throw CrashLogError.readFailed(path: crashPath, underlying: error)
    }
    return FBCrashLog(info: self, contents: contents)
  }

  // MARK: - Predicates

  public class func predicateForCrashLogs(withProcessID processID: pid_t) -> NSPredicate {
    NSPredicate { evaluatedObject, _ in
      guard let crashLog = evaluatedObject as? CrashLogInfo else { return false }
      return crashLog.processIdentifier == processID
    }
  }

  public class func predicateNewer(thanDate date: Date) -> NSPredicate {
    NSPredicate { evaluatedObject, _ in
      guard let crashLog = evaluatedObject as? CrashLogInfo else { return false }
      return date.compare(crashLog.date) == .orderedAscending
    }
  }

  public class func predicateOlder(thanDate date: Date) -> NSPredicate {
    NSCompoundPredicate(notPredicateWithSubpredicate: predicateNewer(thanDate: date))
  }

  public class func predicate(forIdentifier identifier: String) -> NSPredicate {
    NSPredicate { evaluatedObject, _ in
      guard let crashLog = evaluatedObject as? CrashLogInfo else { return false }
      return identifier == crashLog.identifier
    }
  }

  public class func predicate(forName name: String) -> NSPredicate {
    NSPredicate { evaluatedObject, _ in
      guard let crashLog = evaluatedObject as? CrashLogInfo else { return false }
      return name == crashLog.name
    }
  }

  public class func predicate(forExecutablePathContains contains: String) -> NSPredicate {
    NSPredicate { evaluatedObject, _ in
      guard let crashLog = evaluatedObject as? CrashLogInfo else { return false }
      return crashLog.executablePath.contains(contains)
    }
  }

  public class var diagnosticReportsPaths: [String] {
    [
      (NSHomeDirectory() as NSString).appendingPathComponent("Library/Logs/DiagnosticReports"),
      "/Library/Logs/DiagnosticReports",
    ]
  }

  // MARK: - Private

  private class func getPreferredCrashLogParser(forCrashString crashString: String) -> CrashLogParser {
    if !crashString.isEmpty && crashString.first == "{" {
      return ConcatedJSONCrashLogParser()
    } else {
      return PlainTextCrashLogParser()
    }
  }

  private class func fromCrashLogString(_ crashString: String, crashPath: String, parser: CrashLogParser) throws -> CrashLogInfo {
    var executablePath: NSString = NSString()
    var identifier: NSString = NSString()
    var processName: NSString = NSString()
    var parentProcessName: NSString = NSString()
    var processIdentifier: pid_t = -1
    var parentProcessIdentifier: pid_t = -1
    var date: NSDate = NSDate()
    var exceptionDescription: NSString = NSString()
    var crashedThreadDescription: NSString = NSString()

    var parseError: NSError?
    parser.parseCrashLog(
      from: crashString,
      executablePathOut: &executablePath,
      identifierOut: &identifier,
      processNameOut: &processName,
      parentProcessNameOut: &parentProcessName,
      processIdentifierOut: &processIdentifier,
      parentProcessIdentifierOut: &parentProcessIdentifier,
      dateOut: &date,
      exceptionDescription: &exceptionDescription,
      crashedThreadDescription: &crashedThreadDescription,
      error: &parseError
    )

    if let parseError {
      throw CrashLogError.parseFailed(underlying: parseError)
    }

    let processNameStr = processName as String
    if processNameStr.isEmpty {
      throw CrashLogError.missingField(field: "process name")
    }
    let identifierStr = identifier as String
    if identifierStr.isEmpty {
      throw CrashLogError.missingField(field: "identifier")
    }
    let parentProcessNameStr = parentProcessName as String
    if parentProcessNameStr.isEmpty {
      throw CrashLogError.missingField(field: "parent process name")
    }
    let executablePathStr = executablePath as String
    if executablePathStr.isEmpty {
      throw CrashLogError.missingField(field: "executable path")
    }
    if processIdentifier == -1 {
      throw CrashLogError.missingField(field: "process identifier")
    }
    if parentProcessIdentifier == -1 {
      throw CrashLogError.missingField(field: "parent process identifier")
    }

    let processType = self.processType(forExecutablePath: executablePathStr)

    let exceptionDescStr = exceptionDescription as String
    let crashedThreadDescStr = crashedThreadDescription as String

    return CrashLogInfo(
      crashPath: crashPath,
      executablePath: executablePathStr,
      identifier: identifierStr,
      processName: processNameStr,
      processIdentifier: processIdentifier,
      parentProcessName: parentProcessNameStr,
      parentProcessIdentifier: parentProcessIdentifier,
      date: date as Date,
      processType: processType,
      exceptionDescription: exceptionDescStr.isEmpty ? nil : exceptionDescStr,
      crashedThreadDescription: crashedThreadDescStr.isEmpty ? nil : crashedThreadDescStr
    )
  }

  private class func processType(forExecutablePath executablePath: String) -> CrashLogInfoProcessType {
    if executablePath.contains("Platforms/iPhoneSimulator.platform") {
      return .system
    }
    if executablePath.contains(".app") {
      return .application
    }
    return .custom
  }

  private class func predicateForFiles(withBasePath basePath: String, afterDate date: Date, withExtensions extensions: [String]) -> NSPredicate {
    let fileManager = FileManager.default
    let datePredicate: NSPredicate
    datePredicate = NSPredicate { evaluatedObject, _ in
      guard let fileName = evaluatedObject as? String else { return false }
      let path = (basePath as NSString).appendingPathComponent(fileName)
      guard let attributes = try? fileManager.attributesOfItem(atPath: path),
        let modDate = attributes[.modificationDate] as? Date
      else {
        return false
      }
      return modDate.compare(date) != .orderedAscending
    }
    return NSCompoundPredicate(andPredicateWithSubpredicates: [
      NSPredicate(format: "pathExtension in %@", extensions),
      datePredicate,
    ])
  }
}
